Add collection release process (#18)
ansible-collections/cloud-content-handbook: Release/release_collections.md

# Steps to release a collection to Automation hub and Galaxy

1. Determine the collection version based on [Semantic Versioning](https://semver.org/). eg: x.y.z
- For major/breaking change release, increment x.
- For bugfix/minor changes, increment y.
- For patch/trivial changes, increment z.
2. Incase of a major release , create a new branch `stable-X`, where `X` denotes the major version number.

```
git branch stable-X
```
3. For a minor release, selectively backport bug fixes from the main branch to the `stable-X` branch, where X is the major version number, based on the severity of the bug. Avoid backporting new features. Ensure that pull requests intended for backporting are labeled with tags such as `backport-X`, where X refers to the target `stable-X` branch.
4. Check out the `stable-X` branch and prepare the collection for release by following the instructions provided [here](https://docs.ansible.com/ansible/latest/community/collection_contributors/collection_releasing.html#preparing-to-release-a-collection).

```

> git checkout stable-X

Reviewer: I guess the part of creating a prep branch from the stable-x branch is missing !?

# Steps to release a collection to Automation hub and Galaxy

**1.** Determine the collection version based on [Semantic Versioning](https://semver.org/). eg: x.y.z
- For major/breaking change release, increment x.
- For bugfix/minor changes, increment y.
- For patch/trivial changes, increment z.

**2.** Incase of a major release , create a new branch `stable-X`, where `X` denotes the major version number.

> ```
> git branch stable-X

Reviewer 2: ```suggestion
```shell
git checkout main
git checkout -b stable-X main
```
You need to ensure that your main branch is sync with the upstream `main` branch

   ```
   release_summary: < release content >
   ```
   **ii.**  Ensure you have `antsibull-changelog` is installed.
   **iii.** Confirm there are fragments for all known changes in changelogs/fragments.
   **iv.** Run antsibull-changelog release.
   ```
   antsibull-changelog release --version <version>
   ```

**6.** Verify `CHANGELOG.rst` for the presence of all the changelog fragments.

**7.** Commit the changes and push the changes to the upstream repository. This will be the preparation PR for the release.Ensure the CI passes.

**8.** Tag the version in Git and push to GitHub.

```
git tag -m "Release <version>" <version>
git push upstream <version>
```

**9.** Once the [CI for the release](https://ansible.softwarefactory-project.io/zuul/status) passes, notify #ansible-partners about the release, for approval for the release on Automation Hub. To manually upload the collection to Automation Hub refer [here](https://github.com/ansible-collections/cloud-content-handbook/blob/main/Release/release_automation_hub.md).

**10.** Check for the latest version of the collection on Galaxy and Automation Hub.

> **11.** To include an announcment of the release in Bullhorn, add a message in the [Ansible-Social Matric Channel]( https://chat.ansible.im/#/room/#social:ansible.com).

Author: As far as I know, we mention the newsbot on Matrix for each release of any of the cloud collections.

> **11.** To include an announcment of the release in Bullhorn, add a message in the [Ansible-Social Matric Channel]( https://chat.ansible.im/#/room/#social:ansible.com).

Reviewer 3: If tagging newsbot is still required it should be noted here.

# Steps to release a collection to Automation hub and Galaxy

**1.** Determine the collection version based on [Semantic Versioning](https://semver.org/). eg: x.y.z
- For major/breaking change release, increment x. Resulting branch will be x+1.0.0.
- For bugfix/minor changes, increment y. Resulting branch will be x.y+1.0.
- For patch/trivial changes, increment z. Resulting branch will be x.y.z+1.

> **2.** In case of a major release , create a new branch `stable-X`, where `X` denotes the major version number. The new branch is based on the main branch.

Reviewer: ```suggestion
**2.** In case of a major release, create a new branch `stable-X`, where `X` denotes the major version number. The new branch is based on the main branch.
```

```
git checkout main
git branch stable-X main
```

**3.** For a minor release, make sure the backport of the PRs from the main branch to the release branch is successful. For backporting of PRs refer [the document](https://github.com/ansible-collections/cloud-content-handbook/blob/main/backport_changes.md).

**4.** Create and check out a new branch and prepare the collection for release by following the instructions provided [here](https://docs.ansible.com/ansible/latest/community/collection_contributors/collection_releasing.html#preparing-to-release-a-collection).

```
git checkout -b prep_release_x stable-X
```

**5.** Update the CHANGELOG:
  **i.**   Add a changelog fragment `changelogs/fragments/<version>.yml` for the release summary as follows
   ```
   release_summary: < release content >
   ```

>    **ii.**  Ensure you have `antsibull-changelog` is installed.
>    **iii.** Confirm there are fragments for all known changes in changelogs/fragments.
>    **iv.** Run antsibull-changelog release.

Reviewer: ```suggestion
   **ii.** Ensure you have `antsibull-changelog` is installed.
   **iii.** Confirm there are fragments for all known changes in changelogs/fragments.
   **iv.** Run antsibull-changelog release.
```

   ```
   antsibull-changelog release --version <version>
   ```

**6.** Verify `CHANGELOG.rst` for the presence of all the changelog fragments.

> **7.** Commit the changes and push the changes to the upstream repository. Make sure you choose `stable-X` as your base branch. This will be the preparation PR for the release. Ensure the CI passes. Once the `prep` PR is merged, update the local copy of `stable-X` branch with the latest changes from the `prep` PR.

Reviewer: Before committing, we should run the sanity tests locally.

Author: The sanity tests are executed as part of the CI process for the release preparation PR. I shall add a note specifying that in case there are sanity failures observed either in the release preparation PR or locally, a separate PR should be initiated to address those issues.

**8.** Tag the version in Git and push to GitHub.

```
git tag -m "Release <version>" <version>
git push upstream <version>
```

**9.** Once the [CI for the release](https://ansible.softwarefactory-project.io/zuul/status) passes, notify #ansible-partners about the release, for approval for the release on Automation Hub. To manually upload the collection to Automation Hub refer [here](https://github.com/ansible-collections/cloud-content-handbook/blob/main/Release/release_automation_hub.md).

**10.** Check for the latest version of the collection on Galaxy and Automation Hub.

> **11.** To include an announcement of the release in Bullhorn, add a message in the [Ansible-Social Matrix Channel]( https://chat.ansible.im/#/room/#social:ansible.com).

Reviewer: I guess it will be helpful to provide an example template for the message to be posted as well.

> **12.** Create a PR in the collection to merge the changes made in the release PR to the the main branch. Once the CI passes and the PR is approved, merge it to the main branch.

Reviewer: I guess you should provide the steps necessary to do that, like using cherry-pick <sha commit>.

# Steps to release a collection to Automation hub and Galaxy

> **1.** Determine the collection version based on [Semantic Versioning](https://semver.org/). eg: x.y.z

Reviewer 4: A formatting suggestion: let markdown handle the list you. This is especially a problem in step 5 where your nested list is all done inline because markdown does not recognize it as a list.

```
git checkout main
git branch stable-X main
```

**3.** For a minor release, make sure the backport of the PRs from the main branch to the release branch is successful. For backporting of PRs refer [the document](https://github.com/ansible-collections/cloud-content-handbook/blob/main/backport_changes.md).

**4.** Create and check out a new branch and prepare the collection for release by following the instructions provided [here](https://docs.ansible.com/ansible/latest/community/collection_contributors/collection_releasing.html#preparing-to-release-a-collection).

```
git checkout -b prep_release_x stable-X
```

**5.** Update the CHANGELOG:
  **i.** Add a changelog fragment `changelogs/fragments/<version>.yml` for the release summary as follows
   ```
   release_summary: < release content >
   ```
   **ii.** Ensure you have `antsibull-changelog` is installed.
   **iii.** Confirm there are fragments for all known changes in changelogs/fragments.
   **iv.** Run antsibull-changelog release.
   ```
   antsibull-changelog release --version <version>
   ```

**6.** Verify `CHANGELOG.rst` for the presence of all the changelog fragments.

> **7.** Commit the changes and push the changes to the upstream repository. Make sure you choose `stable-X` as your base branch. This will be the preparation PR for the release. Ensure the CI passes. Once the `prep` PR is merged, update the local copy of `stable-X` branch with the latest changes from the `prep` PR.

Reviewer 5: ```suggestion
**7.** Commit the changes and create a pull request into the upstream repository. Make sure you choose `stable-X` as your base branch. This will be the preparation PR for the release. Ensure the CI passes. Once the `prep` PR is merged, update the local copy of `stable-X` branch with the latest changes from the `prep` PR.
```

_Note : If the sanity tests fail locally or in the CI, the failures have to be addressed in a separate PR_

**8.** Tag the version in Git and push to GitHub.

```
git tag -m "Release <version>" <version>
git push upstream <version>
```

> **9.** Once the [CI for the release](https://ansible.softwarefactory-project.io/zuul/status) passes, notify the #ansible-partners slack channel about the release, for approval for the release on Automation Hub. To manually upload the collection to Automation Hub refer [here](https://github.com/ansible-collections/cloud-content-handbook/blob/main/Release/release_automation_hub.md).

Reviewer 4: Is this necessary? My understanding is that they see these releases and do the approval without needing to ask.

Author: I think this aligns with a guideline followed by other teams. I've previously notified the PE team about a release, seeking approval, and I've noticed other teams following a similar practice.

**10.** Check for the latest version of the collection on Galaxy and Automation Hub.

**11.** To include an announcement of the release in Bullhorn, tag the newsbot and add a message in the [Ansible-Social Matrix Channel]( https://chat.ansible.im/#/room/#social:ansible.com).

```
Sample message:
@newsbot
: amazon.aws 7.0.0 has been released with new bugfixes, features, plugins, and modules. Refer the <changelog> for details!
```

**12.** Create a PR in the collection to merge the changes made in the release PR to the the main branch. This can be done manually or using the `cherry-pick` command as given below.
```
_This example assumes that the released version is 3.0.0_
git checkout main
git pull
git checkout -b cherry-pick/stable-3/release_sync
git cherry-pick -x <commitSHA_from_Stable-3>
```
You can also use [CPython’s cherry-picker tool](https://pypi.org/project/cherry_picker/#cherry-picking).

> Push the branch to your fork. Once the CI passes and the PR is approved, merge it to the main branch.

Reviewer 5: ```suggestion
Push the branch to your fork and create a PR into the main branch of the upstream repo. Once the CI passes and the PR is approved, merge it to the main branch.
```

   ```
   git checkout -b prep_release_x stable-X
   ```

* Update the CHANGELOG:
   * Add a changelog fragment `changelogs/fragments/<version>.yml` for the release summary as follows
   ```
   release_summary: < release content >
   ```
   * Ensure you have the latest version of [`antsibull-changelog`](https://ansible.readthedocs.io/projects/antsibull-changelog/) installed.
   * Confirm there are fragments for all known changes in changelogs/fragments.
   * Run antsibull-changelog release.
   ```
   antsibull-changelog release --version <version>
   ```

* Verify `CHANGELOG.rst` for the presence of all the changelog fragments.

* Commit the changes and create a pull request into the upstream repository. Make sure you choose `stable-X` as your base branch. This will be the preparation PR for the release. Ensure the CI passes. Once the `prep` PR is merged, update the local copy of `stable-X` branch with the latest changes from the `prep` PR.
_Note : If the sanity tests fail locally or in the CI, the failures have to be addressed in a separate PR_

* Tag the version in Git and push to GitHub.

   ```
   git tag -m "Release <version>" <version>
   git push upstream <version>
   ```

> * Once the [CI for the release](https://ansible.softwarefactory-project.io/zuul/status) passes, notify the #ansible-partners slack channel about the release, for approval for the release on Automation Hub. To manually upload the collection to Automation Hub refer [here](https://github.com/ansible-collections/cloud-content-handbook/blob/main/Release/release_automation_hub.md).

Reviewer: You should probably mention when the manual upload to the Automation hub should be done.
